src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5682,6 +5682,662 @@ address generate_avx_ghash_processBlocks() {
+      __ jcc(Assembler::notEqual, L_donePadding);
+
+      __ decrementq(r13, 1);
+      __ shrq(rax, 1);
+      __ jmp(L_donePadding);
+
+      __ align(32);
+      __ BIND(L_bruteForce);
+    }   // End of if(avx512_vbmi)
+
+    // Use non-AVX code to decode 4-byte chunks into 3 bytes of output
+
+    // Register state (Linux):
+    // r12-15 - saved on stack
+    // rdi - src
+    // rsi - sp
+    // rdx - sl
+    // rcx - dst
+    // r8 - dp
+    // r9 - isURL
+
+    // Register state (Windows):
+    // r12-15 - saved on stack
+    // rcx - src
+    // rdx - sp
+    // r8 - sl
+    // r9 - dst
+    // r12 - dp
+    // r10 - isURL
+
+    // Registers (common):
+    // length (r14) - bytes in src
+
+    const Register decode_table = r11;
+    const Register out_byte_count = rbx;
+    const Register byte1 = r13;
+    const Register byte2 = r15;
+    const Register byte3 = WINDOWS_ONLY(r12) NOT_WINDOWS(r8);
+    const Register byte4 = WINDOWS_ONLY(rdx) NOT_WINDOWS(rsi);
+
+    __ shrq(length, 2);    // Multiple of 4 bytes only - length is # 4-byte chunks
+    __ cmpq(length, 0);
+    __ jcc(Assembler::lessEqual, L_exit_no_vzero);
+
+    // Set up src and dst pointers properly
+    __ addq(source, start_offset);     // Initial offset
+    __ addq(dest, dp);
+    __ pop(byte2);    // Clear old dest from stack
+    __ push(dest);
+
+    __ shll(isURL, 8);    // index into decode table based on isURL
+    __ lea(decode_table, ExternalAddress(StubRoutines::x86::base64_decoding_table_addr()));
+    __ addq(decode_table, isURL);
+
+    __ jmp(L_bottomLoop);
+
+    __ align(32);
+    __ BIND(L_forceLoop);
+    __ shll(byte1, 18);
+    __ shll(byte2, 12);
+    __ shll(byte3, 6);
+    __ orl(byte1, byte2);
+    __ orl(byte1, byte3);
+    __ orl(byte1, byte4);
+
+    __ incrementq(source, 4);
+
+    __ movb(Address(dest, RegisterOrConstant(), Address::times_1, 2), byte1);
+    __ shrl(byte1, 8);
+    __ movb(Address(dest, RegisterOrConstant(), Address::times_1, 1), byte1);
+    __ shrl(byte1, 8);
+    __ movb(Address(dest, RegisterOrConstant(), Address::times_1, 0), byte1);
+
+    __ incrementq(dest, 3);
+    __ decrementl(length, 1);
+    __ jcc(Assembler::zero, L_exit_no_vzero);
+
+    __ BIND(L_bottomLoop);
+    __ load_signed_byte(byte1, Address(source, RegisterOrConstant(), Address::times_1, 0));
+    __ load_signed_byte(byte2, Address(source, RegisterOrConstant(), Address::times_1, 1));
+    __ load_signed_byte(byte1, Address(decode_table, byte1, Address::times_1, 0));
+    __ load_signed_byte(byte2, Address(decode_table, byte2, Address::times_1, 0));
+    __ load_signed_byte(byte3, Address(source, RegisterOrConstant(), Address::times_1, 2));
+    __ load_signed_byte(byte4, Address(source, RegisterOrConstant(), Address::times_1, 3));
+    __ load_signed_byte(byte3, Address(decode_table, byte3, Address::times_1, 0));
+    __ load_signed_byte(byte4, Address(decode_table, byte4, Address::times_1, 0));
```

Comment:
You could use Address(base, offset) form directly here and other places: e.g. Address (source, 1) instead of Address(source, RegisterOrConstant(), Address::times_1, 1).

Reply:
Done.